util/src/main/java/org/datacommons/util/NameResolver.java

```diff
@@ -0,0 +1,92 @@
+package org.datacommons.util;
+
+import static java.util.stream.Collectors.toList;
+
+import java.util.LinkedHashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.datacommons.proto.Mcf.McfGraph.PropertyValues;
+import org.datacommons.proto.Mcf.McfGraph.TypedValue;
+import org.datacommons.proto.Mcf.McfGraph.Values;
+import org.datacommons.proto.Mcf.ValueType;
+import org.datacommons.proto.Resolve.ResolveRequest;
+import org.datacommons.proto.Resolve.ResolveResponse;
+import org.datacommons.proto.Resolve.ResolveResponse.Entity.Candidate;
+
+/** Resolves nodes with names by calling the DC resolution API. */
+final class NameResolver {
+  // Names to be resolved.
+  private final Set<String> resolveNames = ConcurrentHashMap.newKeySet();
+
+  // Names that were resolved to DCIDs.
+  // The Map maintains mappings from name to Set of candidate DCIDs.
+  private final ConcurrentHashMap<String, Set<String>> resolvedNames = new ConcurrentHashMap<>();
+
+  private final ReconClient client;
+
+  NameResolver(ReconClient client) {
+    this.client = client;
+  }
+
+  boolean submit(PropertyValues node) {
+    Optional<String> optionalName = getName(node);
+    if (optionalName.isPresent()) {
+      resolveNames.add(optionalName.get());
+      return true;
+    }
+    return false;
+  }
+
+  void drain() {
+    if (!resolveNames.isEmpty()) {
+      ResolveRequest request =
+          ResolveRequest.newBuilder()
+              .addAllNodes(resolveNames)
+              .setProperty("<-description->dcid")
+              .build();
+      ResolveResponse response = client.resolve(request);
+      populateResolvedNames(response);
+    }
+  }
+
+  Optional<String> resolve(PropertyValues node) {
+    return getName(node)
+        .filter(resolvedNames::containsKey)
+        .flatMap(name -> resolvedNames.get(name).stream().findFirst());
```

Comment:
Sometimes, many resolved candidates are returned, not sure if we always want to pick first one, or return these to clients

Reply:
This will be called by `ExternalIdResolver` which needs only one value to be returned. But I think what we can do is allow clients to optionally set what type of value they want and resolve based on that. I've added a TODO.